File: Arrays/BuyAndSellStock2/BuyAndSellStock2.go
package main

import (
	"fmt"
	"math"
)

func method1(arr []int) int {
	n := len(arr)
	profit := make([]int, n)
	max := arr[n-1]
	var i int
	for i = n - 2; i >= 0; i-- {
		max = int(math.Max(float64(max), float64(arr[i])))
		profit[i] = int(math.Max(float64(profit[i+1]), float64(max-arr[i])))
	}
	min := arr[0]
	for i = 1; i < n; i++ {
		min = int(math.Min(float64(min), float64(arr[i])))
		profit[i] = int(math.Max(float64(profit[i-1]), float64(profit[i]+arr[i]-min)))
	}
	// fmt.Println("size of arr is ", n)
	// fmt.Print("profit array is ", profit)
	return profit[n-1]
}

func method2(arr []int) int {
	firstBuy := -99999
	firstSell := 0
	secondBuy := -9999
	secondSell := 0
	var i int
	n := len(arr)
	for i = 0; i < n; i++ {
		firstBuy = int((math.Max(float64(firstBuy), float64(-arr[i]))))
		firstSell = int((math.Max(float64(firstSell), float64(firstBuy+arr[i]))))
		secondBuy = int((math.Max(float64(secondBuy), float64(firstSell-arr[i]))))
		secondSell = int((math.Max(float64(secondSell), float64(secondBuy+arr[i]))))
	}
	return secondSell

}

func main() {
	arr := []int{10, 22, 5, 75, 65, 80}
	fmt.Println(method1(arr))
	fmt.Println(method2(arr))
}

File: Arrays/BuyAndSellStock2/BuyAndSellStock2.js
function method1(arr){
    let i, n=arr.length,max=0, min=arr[0];
    let profit = new Array(n);
    //console.log(profit.length);
    profit.fill(0);
    max = arr[n-1];
    for(i=n-2;i>=0;i--){
        max = Math.max(max, arr[i]);
        profit[i] = Math.max(profit[i+1], max-arr[i]);
    }
    //console.log(profit);
    for(i=1;i<n;i++){
        min = Math.min(min, arr[i]);
        profit[i] = Math.max(profit[i-1], profit[i]+(arr[i]-min));
    }
    //console.log('profit is ', profit);
    return profit[n-1];
}

function method2(arr){
    let first_buy = -9999, first_sell = 0, second_buy = -9999, second_sell = 0;
    for(let no of arr){
        first_buy = Math.max(first_buy, -no);
        first_sell = Math.max(first_sell, first_buy+no);
        second_buy = Math.max(second_buy, first_sell-no);
        second_sell = Math.max(second_sell, second_buy+no);
       //console.log(first_buy+","+first_sell+","+second_buy+","+second_sell);
    }
    return second_sell;
}

let arr = [10, 22, 5, 75, 65, 80];
console.log(method1(arr));
console.log(method2(arr));
console.log(method1(arr));
